run_diff.go

```diff
@@ -0,0 +1,142 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package wptdashboard
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strings"
+
+	"golang.org/x/net/context"
+	"google.golang.org/appengine/datastore"
+	"google.golang.org/appengine/urlfetch"
+)
+
+type platformAtRevision struct {
+	Platform string
+	Revision string
+}
+
+func parsePlatformAtRevisionSpec(spec string) (platformAtRevision platformAtRevision, err error) {
+	pieces := strings.Split(spec, "@")
+	if len(pieces) > 2 {
+		return platformAtRevision, errors.New("invalid platform@revision spec: " + spec)
+	}
+	platformAtRevision.Platform = pieces[0]
+	if len(pieces) < 2 {
+		// No @ is assumed to be the platform only.
+		platformAtRevision.Revision = "latest"
+	} else {
+		platformAtRevision.Revision = pieces[1]
+	}
+	if IsBrowserName(platformAtRevision.Platform) {
+		return platformAtRevision, nil
+	}
+	return platformAtRevision, errors.New("Platform " + platformAtRevision.Platform + " not found")
+}
+
+func fetchRunResultsJSONForParam(
+	ctx context.Context, r *http.Request, revision string) (results map[string][]int, err error) {
+	var spec platformAtRevision
+	if spec, err = parsePlatformAtRevisionSpec(revision); err != nil {
+		return nil, err
+	}
+	return fetchRunResultsJSONForSpec(ctx, r, spec)
+}
+
+func fetchRunResultsJSONForSpec(
+	ctx context.Context, r *http.Request, revision platformAtRevision) (results map[string][]int, err error) {
+	var run TestRun
+	if run, err = fetchRunForSpec(ctx, revision); err != nil {
+		return nil, err
+	} else if (run == TestRun{}) {
+		return nil, nil
+	}
+	return fetchRunResultsJSON(ctx, r, run)
+}
+
+func fetchRunForSpec(ctx context.Context, revision platformAtRevision) (TestRun, error) {
+	baseQuery := datastore.
+		NewQuery("TestRun").
+		Order("-CreatedAt").
+		Limit(1)
+
+	var results []TestRun
+	query := baseQuery.
+		Filter("BrowserName =", revision.Platform)
+	if revision.Revision != "latest" {
+		query = query.Filter("Revision = ", revision.Revision)
+	}
+	if _, err := query.GetAll(ctx, &results); err != nil {
+		return TestRun{}, err
+	}
+	if len(results) < 1 {
+		return TestRun{}, nil
+	}
+	return results[0], nil
+}
+
+func fetchRunResultsJSON(ctx context.Context, r *http.Request, run TestRun) (results map[string][]int, err error) {
+	client := urlfetch.Client(ctx)
+	url := run.ResultsURL
+	if strings.Index(url, "/") == 0 {
+		reqURL := *r.URL
+		reqURL.Path = url
+	}
+	var resp *http.Response
+	if resp, err = client.Get(url); err != nil {
```

Comment:
If this is where network errors end up, it seems somewhat likely that we'll see some. Will the client see a HTTP 500 then? If it's not very unlikely to happen, seems like some retry logic will be needed in the Travis integration?

Reply:
Yes, 500, idk about the likelyhood, not going to overengineer until we see an actual problem (travis makes a HUGE number of network fetches for pip etc anyway).

Reply:
Sure, I suppose that this will only make a handful of requests, I was thinking about what @mdittmer told me about in code where he fetched 10k files or so. If we expected that kind of load some precautions might be in order.